[LeonardYam] iP

LeonardYam opened the pull request:
# Bob 😄 
> There are only two hard things in Computer Science: cache invalidation and naming things.

[Phil Karton](https://martinfowler.com/bliki/TwoHardThings.html)

**Bob** is 
- text-based 

He can:
1. List tasks
2. Add tasks
3. Delete tasks
and more...

Bob can help you add a Todo task with the command `todo task`

Current features:
- [x] Managing tasks
- [x] GUI 

---
Here is an example on how to use Bob
```java
public static void main(String[] args) {
        Bob bob = new Bob("data/taskList.txt");
        bob.start();
}
```

Commit:
Level 1 of IP

Commit:
Level 2

Commit:
Added a task list feature to Bob

Commit:
Level-4, added 3 different Tasks

Commit:
Added Gradle and automated testing

Commit:
Added BobException exception class for error handling

Commit:
Added delete feature for task list

Commit:
Add read and write task list to file

Commit:
Integrate LocalDate library

Commit:
Merge branch 'branch-Level-7'
Allows Bob to load a task list from a text file, and save it

Commit:
Merge branch-Level-8 to master

jundatan commented on build.gradle:
```diff
@@ -0,0 +1,24 @@
+        }
+    }
+}
+dependencies {
```
I think it will be better if there is an empty line before this line.

LeonardYam replied:
Done!

jundatan commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+            return false;
+        }
+    }
+    private void handleMarkCommand(String s) throws BobException{
```
Should there be no space after BobException?

LeonardYam replied:
Thanks for catching this!

jundatan reviewed (commented):
Overall coding standards are quite good. I like your naming conventions and they are following the coding standards. There are only a few spacing issues in the code which can be remedy.

TamKengHong commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+import java.nio.file.Path;
+import org.apache.commons.lang3.StringUtils;
+
+public class Bob {
```
Should split into more classes (UI, Storage, TaskList)

huangzz125 replied:
Yup, that will be under A-MoreOOP which I believe Leonard has not committed. Jiayou!

TamKengHong commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+    private final Path dataDirectoryPath = Paths.get("data/");
+    private final Path dataFilePath = Paths.get("data/taskList.txt");
+
+    // Accepts strings that can be separated by \n
```
Should have more javadoc instead of comments

TamKengHong commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    private void readTaskList() {
+        String flag = "bye";
```
can just write: while !input.equals("bye"), dont need a flag variable here

jundatan replied:
I agree. String does not really represent a flag too. So it is better to rename it to something else or use  while !input.equals("bye") 😄

TamKengHong commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    // Command: event <description> /from <start> /to <4pm>
+    private boolean isEvent(String s) {
```
good that you checked for valid input and explained with comments!

TamKengHong commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+
+public class Bob {
+    private final Scanner scanner = new Scanner(System.in);
+
```
inconsistent indentation? Maybe can either put everything with newline or without

LeonardYam replied:
Thanks! I have standardized the spacings

TamKengHong commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+    }
+
+    //Check that the string is a number
+    private static boolean isInt(String s) {
```
why need this? Can try Integer.parseInt() to check if the string is a number, catch when string is not number, dont have to implement int parsing yourself.

LeonardYam replied:
The intention here was to avoid using exceptions as much as possible

TamKengHong reviewed (commented):
May need a little bit of refactoring + javadoc
Otherwise the code is self explanatory and good variable naming.

huangzz125 commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.Arrays;
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.Path;
+import org.apache.commons.lang3.StringUtils;
```
The imports should be in lexicographical order.

huangzz125 commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+        String[] splitCommand = s.split(" /from | /to ");
+
+        // Check if a description exists
+        String[] command_desc = splitCommand[0].split(" ");
```
Should be named as commandDesc instead.

huangzz125 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+    }
+    public String toString() {
+        DateTimeFormatter f = DateTimeFormatter.ofPattern("MMM dd yyyy");
+        return String.format("%s | by: %s", this.description, this.deadline.format(f));
```
Good use of String.format() method!

huangzz125 reviewed (commented):
In general, I think most of the coding standard requirements are adhered to since most methods and variables are adequately named. Since you have not completed the A-Checkstyle yet, it is understandable that your styling has not adjusted according to the checkstyle standards. Other than that, great work!

jundatan commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+    private final Scanner scanner = new Scanner(System.in);
+
+    private ArrayList<Task> taskList = new ArrayList<>();
+    private static final Integer spacing = 5;
```
Should the static variables be separated from each other?

jundatan commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+                && isInt(words[1]);
+    }
+
+    // Command: todo <description>
```
I think the method commands are not following the coding standard.

jundatan commented on src/main/java/Bob.java:
```diff
@@ -0,0 +1,338 @@
+    /**
+     * Main program for Bob, our chat-bot
+     */
+    public void start() {
+        //Introduction message
+        formattedPrint("Hi, my name is Bob :)\n" +
+                "How may I help you?");
+
+        // Load task list from file
+        loadList();
+
+        // Read tasks from user
+        readTaskList();
+
+        // Save task list to file
+        saveList();
+
+        // Goodbye message
+        formattedPrint("See you soon!");
+    }
+}
```
Good and tidy naming conventions and comments made here 👍 Easily readable.

jundatan reviewed (commented):
Overall good in terms of coding standards as stated in my previous review. Wanted to add more comments so i am making a new review. Only major issue is missing/incorrect header comments for most methods and classes.

LeonardYam committed:
Add A-MoreOOP increment

LeonardYam committed:
Create Bob package

LeonardYam committed:
Add JUnit tests

LeonardYam committed:
Add JavaDocs

LeonardYam committed:
Implement Coding Standard

LeonardYam committed:
Add find feature for tasks

LeonardYam committed:
Merge branch 'branch-A-JavaDoc'

LeonardYam committed:
Merge branch 'branch-A-CodingStandard'

LeonardYam committed:
Merge branch 'branch-Level-9'

LeonardYam committed:
Add JavaFX GUI

LeonardYam committed:
Refactor JavaFX

LeonardYam committed:
Add assertions to various methods

Several methods have assumptions about certain preconditions
and preconditions.

Adding assertions to the code help verify these assumptions and makes
error handling easier.

LeonardYam committed:
Merge pull request #2 from LeonardYam/branch-A-Assertions

Add assertions to various methods

LeonardYam committed:
Integrated Checkstyle to detect style violations

LeonardYam committed:
Merge branch 'branch-A-CheckStyle'

LeonardYam committed:
Refactor code to improve code quality

Several methods were too long and not readable.

Refactoring these methods improves code quality and readability.

LeonardYam committed:
Merge pull request #3 from LeonardYam/branch-A-CodeQuality

Refactor code to improve code quality

Commit:
Add B-Reminders extension

Currently, users can list their tasks
but aren't able to see their upcoming deadlines.

Adding a reminder feature helps users
keep track of upcoming deadlines.

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Add more JUnit tests

More JUnit tests helps with regression testing
and ensures Bob is working as expected.

LeonardYam committed:
Improve the GUI

Added some padding and colours to make the
GUI interface for Bob more aesthetic.

LeonardYam committed:
Update user guide

LeonardYam committed:
Add a complete user guide

LeonardYam committed:
Added more JavaDocs

LeonardYam committed:
Add Regex and add JavaDocs

LeonardYam committed:
Update build.gradle file

LeonardYam committed:
Upload Ui.png

LeonardYam committed:
Add Jekyll configuration file for Pages

LeonardYam committed:
Fix user guide

LeonardYam committed:
Update _config.yml

LeonardYam committed:
Update README.md
